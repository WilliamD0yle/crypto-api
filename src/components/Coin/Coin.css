.coin-container {
    display: flex;
    flex: 1;
    flex-direction: column;
    justify-content: center;
    align-items: center;
    width: 80%;
    margin-left: 10%;
    margin-right: 10%;
    background-color: rgba(255, 255, 255, 0.644);
}

.coin-header h1 {
    color: rgb(9, 17, 63);
}

.row-header {
    display: flex;
    flex-direction: row;
}

.row-header {
    display: flex;
    flex-direction: row;
    justify-content: space-around;
    width: 100%;
    border-radius: 5px;
    background-color: rgb(9, 17, 63);
}

.row-header p {
    cursor: pointer;
    color: rgb(255, 255, 255);
}

.row-header p:hover {
    color: rgb(22, 41, 94);
    text-decoration: underline;
}

.coin-details {
    display: flex;
    flex-direction: row;
    justify-content: center;
    width: 100%;
}

.coin-column {
    display: flex;
    flex: 1;
    flex-direction: column;
    justify-content: center;
    align-items: flex-start;
}
.coin-column img {
    display: flex;
    justify-self: center;
    align-self: center;
}