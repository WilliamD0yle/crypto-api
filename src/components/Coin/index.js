import React, { useEffect, useState } from 'react';
import { connect } from 'react-redux';
import Graphs from '../Graphs';
import { getCoinDetails, getHistoricDetails } from '../../actions/';
import './Coin.css';
const mediaUrl = 'https://www.cryptocompare.com';

export const Coin = props => {
    let [data, updateData] = useState([]);
    let [selectedGraph, updateGraph] = useState('line');
    let coin = props.match.params.coin;

    useEffect(() => {
        if (!props.details) {
            props.getCoinDetails(coin);
        }

        let graphData = props[props.selected];
        if (graphData.length > 0) {
            let data = graphData.map((obj, i) => ({
                x: obj.time,
                y: obj.close
            }));
            updateData(data);
        }
        // eslint-disable-next-line react-hooks/exhaustive-deps
    }, [props]);

    const renderRow = items => {
        return (
            <div className="row-header">
                {items.map((item, i) => {
                    return (
                        <p key={i} className={item.name} onClick={item.onClick}>
                            {item.name}
                        </p>
                    );
                })}
            </div>
        );
    };

    return (
        <div className="coin-container">
            <div className="coin-header">
                <h1>{props.match.params.coin} Breakdown</h1>
            </div>
            {renderRow([
                { name: 'Line', onClick: () => updateGraph('line') },
                { name: 'Bar', onClick: () => updateGraph('bar') },
                { name: 'Area', onClick: () => updateGraph('area') }
            ])}
            <Graphs
                type={selectedGraph}
                xTitle={'Hour'}
                yTitle={'Price'}
                data={data}
            />
            {renderRow([
                {
                    name: '1 hour',
                    onClick: () =>
                        props.getHistoricDetails(
                            'hour',
                            coin,
                            'histominute',
                            60
                        )
                },
                {
                    name: '24 hrs',
                    onClick: () =>
                        props.getHistoricDetails('day', coin, 'histohour', 24)
                },
                {
                    name: '1 week',
                    onClick: () =>
                        props.getHistoricDetails('week', coin, 'histohour', 168)
                },
                {
                    name: '1 month',
                    onClick: () =>
                        props.getHistoricDetails('month', coin, 'histoday', 30)
                },
                {
                    name: '6 months',
                    onClick: () =>
                        props.getHistoricDetails(
                            'sixMonths',
                            coin,
                            'histoday',
                            180
                        )
                },
                {
                    name: '1 year',
                    onClick: () =>
                        props.getHistoricDetails('year', coin, 'histoday', 365)
                }
            ])}

            {props.details ? (
                <div className="coin-details">
                    <div className="coin-column">
                        <img
                            src={`${mediaUrl}${props.details.IMAGEURL}`}
                            alt={coin}
                        />
                    </div>
                    <div className="coin-column">
                        <p>
                            1 {coin} = {props.details.PRICE}
                        </p>
                        <p>Last trade id: {props.details.LASTTRADEID}</p>
                        <p>
                            Vol Traded 24hrs: {props.details.VOLUME24HOUR} (
                            {props.details.VOLUME24HOURTO})
                        </p>
                        <p>High hr: {props.details.HIGHHOUR}</p>
                        <p>Low hr: {props.details.LOWHOUR}</p>
                        <p>High 24hr: {props.details.HIGH24HOUR}</p>
                        <p>Low 24hr: {props.details.LOW24HOUR}</p>
                    </div>
                </div>
            ) : null}
        </div>
    );
};

const mapStateToProps = state => ({
    details: state.crypto.coin.details,
    selected: state.crypto.coin.selected,
    hour: state.crypto.coin.hour,
    day: state.crypto.coin.day,
    week: state.crypto.coin.week,
    month: state.crypto.coin.month,
    sixMonths: state.crypto.coin.sixMonths,
    year: state.crypto.coin.year
});

const mapDispatchToProps = dispatch => ({
    getCoinDetails: coin => dispatch(getCoinDetails(coin)),
    getHistoricDetails: (key, coin, option, limit) =>
        dispatch(getHistoricDetails(key, coin, option, limit))
});

export default connect(
    mapStateToProps,
    mapDispatchToProps
)(Coin);
