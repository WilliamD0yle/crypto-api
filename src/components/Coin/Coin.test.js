import React from 'react';
import { configure, shallow } from 'enzyme';
import { StaticRouter } from 'react-router';
import Adapter from 'enzyme-adapter-react-16';
import Coin from './';
configure({ adapter: new Adapter() });

describe('Coin component', () => {
    it('renders coin screen without crashing', () => {
        const match = { params: { coin: 'LTC' } };
        const details = {
            IMAGEURL: 'https://www.cryptocompare.com/media/19633/btc.png',
            PRICE: '£ 8,487.97',
            LASTTRADEID: '7020975',
            VOLUME24HOUR: 'Ƀ 336.17',
            VOLUME24HOURTO: '£ 2,852,913.4',
            HIGHHOUR: '£ 8,511.49',
            LOWHOUR: '£ 8,464.04',
            HIGH24HOUR: '£ 8,624.47',
            LOW24HOUR: '£ 8,335.43'
        };
        const coin = shallow(
            <StaticRouter>
                <Coin match={match} details={details} />
            </StaticRouter>
        );
        expect(coin).toMatchSnapshot();
    });
});
