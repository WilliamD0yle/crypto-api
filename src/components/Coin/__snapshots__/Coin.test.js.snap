// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Coin component renders coin screen without crashing 1`] = `ShallowWrapper {}`;
